File: python/interpreter/scanner.py
from error_reporter import ErrorReporter
from lox_token import Token
from token_type import TokenType, get_keyword


class Scanner:
    def __init__(self, source: str, error_reporter: ErrorReporter):
        self.source = source
        self.error_reporter: ErrorReporter = error_reporter
        self.tokens: list[Token] = []
        self.line: int = 1
        self.start: int = 0
        self.current: int = 0
        return

    def scan_tokens(self) -> list[Token]:
        while not self.at_end():
            self.start = self.current
            self.scan_token()

        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
        return self.tokens

    def at_end(self) -> bool:
        return self.current >= len(self.source)

    def scan_token(self):
        c = self.advance()
        match c:
            case '(':
                self.add_token(TokenType.LEFT_PAREN)
            case ')':
                self.add_token(TokenType.RIGHT_PAREN)
            case '{':
                self.add_token(TokenType.LEFT_BRACE)
            case '}':
                self.add_token(TokenType.RIGHT_BRACE)
            case ',':
                self.add_token(TokenType.COMMA)
            case '.':
                self.add_token(TokenType.DOT)
            case '-':
                self.add_token(TokenType.MINUS)
            case '+':
                self.add_token(TokenType.PLUS)
            case ';':
                self.add_token(TokenType.SEMICOLON)
            case '*':
                self.add_token(TokenType.STAR)
            case '=':
                self.add_token(TokenType.EQUAL_EQUAL) if self.match('=') else self.add_token(TokenType.EQUAL)
            case '!':
                self.add_token(TokenType.BANG_EQUAL) if self.match('=') else self.add_token(TokenType.BANG)
            case '>':
                self.add_token(TokenType.GREATER_EQUAL) if self.match('=') else self.add_token(TokenType.GREATER)
            case '<':
                self.add_token(TokenType.LESS_EQUAL) if self.match('=') else self.add_token(TokenType.LESS)
            case '/':
                if self.match('/'):
                    self.add_comment()
                else:
                    self.add_token(TokenType.SLASH)
            case ' ' | '\r' | '\t':
                pass
            case '"':
                self.add_string()
            case c if self.is_digit(c):
                self.add_number()
            case c if self.is_alpha(c):
                self.add_identifier()
            case '\n':
                self.line += 1
            case _:
                self.error_reporter.error(self.line, "Unexpected character")

    def is_digit(self, c: str) -> bool:
        return '0' <= c <= '9'

    def is_alpha(self, c: str) -> bool:
        return ('a' <= c <= 'z') or ('A' <= c <= 'Z') or c == '_'

    def add_token(self, token_type: TokenType, literal=None):
        text = self.source[self.start:self.current]
        self.tokens.append(Token(token_type, text, literal, self.line))

    def add_comment(self):
        while (peek := self.peek()) not in ('/0', "\n"):
            self.advance()

    def add_string(self):
        while (peek := self.peek()) not in ('/0', '"'):
            self.advance()
            if peek == '\n':
                self.line += 1

        if self.peek() == "/0":
            self.error_reporter.error(self.line, "Unterminated string")
            return

        self.advance()
        self.add_token(TokenType.STRING, self.source[self.start + 1:self.current - 1])

    def add_number(self):
        while self.is_digit(self.peek()):
            self.advance()

        if self.peek() == "." and self.is_digit(self.peek_next()):
            self.advance()
            while self.is_digit(self.peek()):
                self.advance()

        self.add_token(TokenType.NUMBER, self.source[self.start:self.current])

    def add_identifier(self):
        while (peek := self.peek()) and self.is_alpha(peek) or self.is_digit(peek):
            self.advance()

        identifier = self.source[self.start:self.current]
        if keyword := get_keyword(identifier):
            self.add_token(keyword, identifier)
        else:
            self.add_token(TokenType.IDENTIFIER, identifier)

    # check if the next character matches, if so consume it and return True else return False
    def match(self, next_c: str) -> bool:
        peek = self.peek()
        if peek == '/0' or peek != next_c:
            return False

        self.advance()
        return True

    # consume a character
    def advance(self) -> str:
        c = self.source[self.current]
        self.current += 1
        return c

    # check next character, but don't consume
    def peek(self) -> str:
        return self.source[self.current] if not self.at_end() else '/0'

    # 2 character lookahead
    def peek_next(self) -> str:
        return self.source[self.current + 1] if not self.current + 1 >= len(self.source) else '/0'

File: tslox/interpreter/scanner.ts
import { reportError } from "./errorReporter";
import { Token } from "./token";
import { TokenType, KeywordToTokenType } from "./tokenType";

export class Scanner {
  private start: number = 0;
  private current: number = 0;
  private line: number = 1;
  private tokens: Token[] = [];

  constructor(private readonly source: string) {}
  scanTokens(): Token[] {
    while (!this.isAtEnd()) {
      this.start = this.current;
      this.scanToken();
    }

    this.tokens.push(new Token(TokenType.EOF, "", null, this.line));
    return this.tokens;
  }

  private isAtEnd(): Boolean {
    return this.current >= this.source.length;
  }

  private scanToken() {
    let c = this.advance();
    switch (c) {
      // Single character
      case "(": {
        this.extractToken(TokenType.LEFT_PAREN);
        break;
      }
      case ")": {
        this.extractToken(TokenType.RIGHT_PAREN);
        break;
      }
      case "{": {
        this.extractToken(TokenType.LEFT_BRACE);
        break;
      }
      case "}": {
        this.extractToken(TokenType.RIGHT_BRACE);
        break;
      }
      case ",": {
        this.extractToken(TokenType.COMMA);
        break;
      }
      case ".": {
        this.extractToken(TokenType.DOT);
        break;
      }
      case "-": {
        this.extractToken(TokenType.MINUS);
        break;
      }
      case "+": {
        this.extractToken(TokenType.PLUS);
        break;
      }
      case ";": {
        this.extractToken(TokenType.SEMICOLON);
        break;
      }
      case "*": {
        this.extractToken(TokenType.STAR);
        break;
      }

      // double characters
      case "!": {
        this.extractToken(
          this.match("=") ? TokenType.BANG_EQUAL : TokenType.BANG
        );
        break;
      }
      case "=": {
        this.extractToken(
          this.match("=") ? TokenType.EQUAL_EQUAL : TokenType.EQUAL
        );
        break;
      }
      case "<": {
        this.extractToken(
          this.match("=") ? TokenType.LESS_EQUAL : TokenType.LESS
        );
        break;
      }
      case ">": {
        this.extractToken(
          this.match("=") ? TokenType.GREATER_EQUAL : TokenType.GREATER
        );
        break;
      }
      case "/": {
        if (this.match("/")) {
          this.extractComment();
        } else {
          this.extractToken(TokenType.SLASH);
        }
        break;
      }
      case " ":
      case "\r":
      case "\t":
        break;

      case "\n":
        this.line++;
        break;
      case '"':
        this.extractString();
        break;
      default: {
        if (this.isDigit(c)) {
          this.extractNumber();
        } else if (this.isAlpha(c)){
          this.extractIdentifier()
        }
        else {
          reportError(this.line, "Unexpected character.");
        }
      }
    }
  }

  private advance(): string {
    return this.source[this.current++];
  }

  private peek(n: number = 0): string {
    if (this.isAtEnd()) return "\0";
    return this.source[this.current + n];
  }

  private extractToken(type: TokenType, literal: string | number = null) {
    let token = new Token(
      type,
      this.source.substring(this.start, this.current),
      literal,
      this.line
    );
    this.tokens.push(token);
  }

  private match(next: string): Boolean {
    if (this.isAtEnd()) {
      return false;
    }
    if (this.source[this.current] != next) {
      return false;
    }

    this.current++;
    return true;
  }

  private extractComment() {
    let next = this.peek();
    while (next != "\n" && next != "\0") {
      this.advance();
      next = this.peek();
    }
  }

  private extractString() {
    let next = this.peek();
    while (next != '"' && next != "\0") {
      if (next == "\n") {
        this.line++;
      }
      this.advance();
      next = this.peek();
    }

    if (next == "\0") {
      reportError(this.line, "Unterminated string");
      return;
    }

    this.advance();

    let s = this.source.substring(this.start + 1, this.current - 1);
    this.extractToken(TokenType.STRING, s);
  }

  private isDigit(c: string): Boolean {
    return c >= "0" && c <= "9";
  }

  private isAlpha(c: string): boolean {
    return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c == "_";
  }

  private extractNumber() {
    let next = this.peek();
    while (this.isDigit(next) || (next == "." && this.isDigit(this.peek(1)))) {
      this.advance();
      next = this.peek();
    }

    if (next == "." && !this.isDigit(this.peek(1))) {
      // consume the offending period
      this.advance();
      reportError(this.line, "Not a valid number");
      return;
    }

    let n = this.source.substring(this.start, this.current);
    this.extractToken(TokenType.NUMBER, parseFloat(n));
  }

  private extractIdentifier() {
    let next = this.peek();
    while(this.isAlpha(next) || this.isDigit(next)) {
      this.advance();
      next = this.peek();
    }

    const identifier = this.source.substring(this.start, this.current);
    let type = TokenType.IDENTIFIER
    if (KeywordToTokenType.hasOwnProperty(identifier)) {
      type = KeywordToTokenType[identifier];
    }
    this.extractToken(type);
  }
}
